File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans = []
        stack = []

        def combination (index, sum, target, stack, nums):
            if sum == target:
                ans.append(stack[:])
                return

            if sum > target or index > len(nums) - 1:
                return
            
            stack.append(nums[index])
            combination(index, sum + nums[index], target, stack, nums)
            stack.pop()
            combination(index + 1, sum, target, stack, nums)

        combination(0, 0, target, stack, nums)
        return ans

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int length = nums.length;
        int[] ans = new int[length];

        ans[length - 1] = nums[length - 1];
        for(int i = nums.length - 2; i >= 0; i--){
            ans[i] = ans[i + 1] * nums[i];
        }

        for(int i = 1; i < length; i++){
            nums[i] *= nums[i - 1];
        }

        for(int i = 0; i < length; i++){
            int pre = 1;
            int post = 1;

            if (i - 1 >= 0)
                pre = nums[i - 1];
            if (i + 1 < length)
                post = ans[i + 1];
            
            ans[i] = pre * post;
        }

        return ans;
    }
}  
